flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDataDeserializationSchema.java

```diff
@@ -106,7 +110,79 @@ public Builder(
+
+            if (rowResultTypeFields.size() < rowReadTypeFields.size()) {
+                // Some fields are filtered from the results.
+                final List<RowType.RowField> fields = rowReadType.getFields();
+
+                final Collection<String> filteredFieldsNames =
+                        getFilteredFieldsNames(rowReadTypeFields, rowResultTypeFields);
+
+                final Set<Integer> filteredFieldsIndices =
+                        filteredFieldsNames.stream()
+                                .map(rowReadType::getFieldIndex)
+                                .collect(Collectors.toSet());
+
+                // This type corresponds to the CsvSchema.ColumnType.STRING in the converter
+                final VarCharType varCharType = new VarCharType();
+                final List<RowType.RowField> optimizedRowFields =
+                        fields.stream()
+                                .map(
+                                        field -> {
+                                            int fieldIndex =
+                                                    rowReadType.getFieldIndex(field.getName());
+                                            if (filteredFieldsIndices.contains(fieldIndex)) {
+                                                return new RowType.RowField(
+                                                        field.getName(),
+                                                        varCharType,
+                                                        field.getDescription().orElse(null));
+                                            } else {
+                                                return field;
+                                            }
+                                        })
+                                .collect(Collectors.toList());
+                return new RowType(rowReadType.isNullable(), optimizedRowFields);
+            } else {
+                return rowReadType;
+            }
+        }
+
+        /**
+         * Detects fields that need to be filtered from the produced rows (for instance, due to a
+         * projection pushdown).
+         */
+        private static Collection<String> getFilteredFieldsNames(
+                List<String> rowReadTypeFields, List<String> rowResultTypeFields) {
+
+            final Collection<String> diff = new ArrayList<>();
+
+            int i = 0, j = 0;
+            while (i < rowReadTypeFields.size()) {
+                final String rowReadFieldName = rowReadTypeFields.get(i);
+                final String rowResultFieldName = rowResultTypeFields.get(j);
+                if (rowReadFieldName.equals(rowResultFieldName)) {
+                    i++;
+                    j++;
+                } else {
+                    // This field is filtered from the results
+                    diff.add(rowReadTypeFields.get(i));
+                    i++;
+                }
+            }
+            return diff;
         }
```

Comment:
Instead of reverse engineering the projection, why not pass Projection to the ctor?